tests/fe/common/fe_shape_function_plotter.h

```diff
@@ -0,0 +1,309 @@
+  }
+  virtual ~FEExtractorFunction() = default;
+
+  virtual void
+  vector_value(const Point<dim> &p, Vector<double> &values) const override
+  {
+    Assert(values.size() == this->n_components, ExcInternalError());
+    Assert(GeometryInfo<dim>::is_inside_unit_cell(p), ExcInternalError());
+
+    // For elements that don't support component-wise extraction, we use
+    // the  FEValuesExtractors / FEValuesViews in conjunction with a
+    // custom quadrature rule on the unit cell.
+    const Quadrature<dim> quad(p);
+    FEValues<dim>         fe_values(fe, quad, update_values);
+    Assert(fe_values.get_quadrature().size() == 1, ExcInternalError());
+
+    Assert(tr.n_active_cells() == 1, ExcInternalError());
+    fe_values.reinit(dof.begin_active());
+    constexpr unsigned int q_point = 0;
+
+    if (this->n_components == 1)
+      {
+        const double Nx = fe_values[fe_s].value(bf, q_point);
+        values[0]       = Nx;
+      }
+    else if (this->n_components == dim)
+      {
+        const Tensor<1, dim> Nx = fe_values[fe_v].value(bf, q_point);
+        for (unsigned int i = 0; i < this->n_components; ++i)
+          values[i] = Nx[i];
+      }
+    else
+      {
+        AssertThrow(false, ExcNotImplemented());
+      }
+  }
+
+private:
+  const FE_Type &    fe;
+  const unsigned int bf;
+
+  Triangulation<dim>         tr;
+  DoFHandler<dim>            dof;
+  FEValuesExtractors::Scalar fe_s;
+  FEValuesExtractors::Vector fe_v;
+};
+
+
+/**
+ * The main function that is used to plot finite element shape
+ * functions. They are output in VTU format, so that they can
+ * be visualised in Paraview and Visit.
+ */
+template <typename FE_Type>
+void
+plot_FE_shape_functions(const std::string    filename,
+                        const unsigned int   order,
+                        const ReferenceCell &reference_cell,
+                        const unsigned int   n_div)
+{
+  const unsigned int dim = FE_Type::dimension;
+
+  // Vector valued finite element to be visualised
+  const FE_Type      fe_viz(order);
+  const unsigned int n_components = fe_viz.n_components();
+  Assert((n_components == 1) || (n_components == dim), ExcInternalError());
+
+  // FESystem for the vector-valued FE to be visualised
+  const bool is_hyper_cube = reference_cell.is_hyper_cube();
+  const bool is_simplex    = reference_cell.is_simplex();
+  const bool is_pyramid    = (reference_cell == ReferenceCells::Pyramid);
+  const bool is_wedge      = (reference_cell == ReferenceCells::Wedge);
+  if (!is_hyper_cube && !is_simplex && !is_pyramid)
+    {
+      (void)is_wedge;
+      Assert(is_wedge, ExcInternalError());
+    }
+
+  const FESystem<dim> fe(
+    is_hyper_cube ?
+      FESystem<dim>(FE_Q<dim>(1), n_components) :
+      (is_simplex ?
+         FESystem<dim>(FE_SimplexP<dim>(1), n_components) :
+         (is_pyramid ? FESystem<dim>(FE_PyramidP<dim>(1), n_components) :
+                       FESystem<dim>(FE_WedgeP<dim>(1), n_components))));
```

Comment:
It took me a while to understand what this is for. Would you mind adding a comment of the form "Define a linear finite element with as many vector components as the target element has, and then interpolate the target element onto this element (on a mesh that is a refined version of the reference cell)."